text/1009-group-by.rst

```diff
@@ -0,0 +1,394 @@
+perform aggregation.
+
+
+Motivation
+==========
+
+Aggregation of data by keys is a key operation in analytics-like
+queries, and EdgeQL currently serves poorly.
+
+Consider trying to, in our "cards" test basebase representing a
+collectible card game, finding the average cost of a card for each of
+the different "elements" are card could have. In EdgeQL now, this can
+be expressed as::
+
+  FOR element in (DISTINCT Card.element) UNION (
+      WITH g := (SELECT Card FILTER .element = element)
+      SELECT {
+          element := element,
+          avg_cost := sum(g.cost) / count(g)
+      } ORDER BY .element
+  )
+
+This query, while it gets the job done is pretty ugly and
+distressingly manual. On top of that, it generates a bad query that
+needs to do a subquery scan for each element!
+
+Implementing a ``GROUP`` statement in EdgeQL will both allow us to
+provide much better developer experience when doing aggregating
+queries and give us the structure needed to target Postgres's
+aggregation features in compilation.
+
+
+Specification
+=============
+
+The new ``GROUP ... BY`` proposal is an objects-forwards proposal,
+which produces a set of free objects containing each produced
+group. The encouraged idiom will be to immediately process the results
+with a ``SELECT`` statement containing a shape that performs
+aggregation, in most cases.
+
+::
+
+   GROUP
+       [<alias> := ] <expr>
+
+   BY
+       [<alias1> := ] <expr>,  # define parameters to use for
+       [<alias2> := ] <expr>,  # grouping
+       ...
+       [<aliasN> := ] <expr>
+
+   [USING <alias>, ... ]       # specify which parameters will
+                               # be used to partition the set;
+                               # if unspecified, use all aliases declared in BY
+
+
+
+The aliases in the BY clause are optional only if the expression is an
+partial path reference such as ``.name``, and ``.name`` is treated as
+equivalent to ``name := .name``, except that ``name`` is not made available
+to later aliases in the ``BY`` clause.
+
+The ``USING`` clause, if omitted, is taken to be the list of all specified
+aliases in the ``BY`` clause. This means it is usually superfluous in this
+initial pre-grouping sets version, but it does allow aliases in the BY
+clause to be defined solely as a helper to later aliases.
+
+The optional alias in the subject will be bound in the ``BY`` clauses.
+
+The *behavior* of ``GROUP BY`` is to evaluate the subject expression and
+then, for each element of the set, evaluate each of the ``BY``
+expressions (which much be singletons). Then a grouping key is created
+based on the ``USING`` clause, and the objects are aggregated into groups
+with matching grouping keys.
+
+The output is reflected in an ad-hoc free object of the following form::
+
+  {
+      key: <another free object containing "alias := value" pointers>,
+      elements: <the set of all the objects with the matching key>,
+      grouping: <array of keys used in grouping>,
```

Comment:
Does it have to be an array?

Reply:
Array is the most convenient representation.  It's actually fairly hard to compare sets in EdgeQL currently.

Reply:
Will you need to compare groupings? There's `array_agg` if you want an array.

Reply:
Comparing them is pretty much the only thing you would do.  `if grouping = [foo, bar]`.  A set would also be more expensive, because internally it's going to be serialized as an array anyway, so you'd be doing pointless `array_agg(unnest(`.  Why do you think a set is more appropriate here?

Reply:
I'm OK with keeping it an array then, thanks for the explanation.

I was wondering if we could use a set here because using an array *feels* a bit weird. Most of our clauses produce sets, and set is the most fundamental "bunch of elements" data type in EdgeQL. If, however, there are good reasons for keeping it an array I'm fine with that.

Reply:
The only argument against it being an array that I can think of is that the ordering is not apparent (is it alphabetic or is it from the order of keys in `GROUPINGS`?)  A set would naturally not have this problem because you'd have to specify the order explicitly.

Reply:
> The only argument against it being an array that I can think of is that the ordering is not apparent (is it alphabetic or is it from the order of keys in GROUPINGS?

It must be from the order of keys in GROUPINGS/USING or it will be unusable. Your own example `if grouping = [foo, bar]` only makes sense if the order matches the order of keys.

Reply:
Yes, of course.  I meant that this might not be apparent to the user, but I may be overthinking it.

Reply:
Likely you are; I don't think it's a problem.

Reply:
> It must be from the order of keys in GROUPINGS/USING or it will be unusable.

I specified it as being based on the order of binding in the `BY` clause. `USING` doesn't work exactly because a key could show up multiple times, like `(b, a), (a, b, c)`, and it would mean we can generate code to produce `grouping` that has one clause per key, not one per grouping.

Doing it based on the order from USING is also doable, though. If we have something like `(a, b), (b, a)` (probably because of CUBE/ROLLUP), we'd just have to pick one.

(I should also update the spec to indicate that if a group appears multiple times in USING, it still only gets output once. (Since that is what SQL does...))